clippy_utils/src/lib.rs

```diff
@@ -707,18 +710,25 @@ pub fn def_path_res(tcx: TyCtxt<'_>, path: &[&str]) -> Vec<Res> {
         None
     };
 
-    let starts = find_primitive_impls(tcx, base)
-        .chain(find_crates(tcx, base_sym))
+    let crates = find_primitive_impls(tcx, base)
         .chain(local_crate)
-        .map(|id| Res::Def(tcx.def_kind(id), id));
+        .map(|id| Res::Def(tcx.def_kind(id), id))
+        .chain(find_crates(tcx, base_sym))
+        .collect();
 
-    let mut resolutions: Vec<Res> = starts.collect();
+    def_path_res_in_crates(tcx, crates, path)
+}
 
+/// Resolves a def path like `vec::Vec` with the crate `std`.
+///
+/// This is lighter than [`def_path_res`], and should be called with [`find_crates`] looking up
+/// items from the same crate repeatedly, although should still be used sparingly.
+pub fn def_path_res_in_crates(tcx: TyCtxt<'_>, mut crates: Vec<Res>, mut path: &[&str]) -> Vec<Res> {
```

Comment:
I think a better name would be something like `def_path_res_with_base`, because the resolutions here aren't always crates. The `Vec` can also include a primitive impl when calling `def_path_res(&["str", "len"])`, and this function looks like it could really work with any mod/impl resolution

Reply:
Done.